File: IntelliJ Projects/CreateShapes/src/SpotColorShapes.java
/* -----------------------------------------------------------------------
 * <copyright file="SpotColorShapes.java" company="Hybrid Software Helix Ltd">
 *  Copyright (C) 2025 Hybrid Software Helix Ltd. All rights reserved.
 * </copyright>
 * <summary>
 *  This example is provided on an "as is" basis and without warranty of any kind.
 *  Hybrid Software Helix Ltd. does not warrant or make any representations
 *  regarding the use or results of use of this example.
 * </summary>
 * -----------------------------------------------------------------------
 */

import com.globalgraphics.JawsMako.jawsmakoIF.*;
import java.lang.Math;

public class SpotColorShapes {

    enum ShapeType { Box, Ellipse, Hexagon, Polygon, Target }

    public static void main(String[] args) {
        try {
            var mako = IJawsMako.create();
            IJawsMako.enableAllFeatures(mako);
            var factory = mako.getFactory();

            var assembly = IDocumentAssembly.create(mako);
            var document = IDocument.create(mako);
            assembly.appendDocument(document);
            var page = IPage.create(mako);
            document.appendPage(page);
            var fixedPage = IDOMFixedPage.create(factory);
            page.setContent(fixedPage);

            // Create some process color brushes
            var cyan = IDOMSolidColorBrush.createSolidCmyk(factory, 1.0f, 0.0f, 0.0f, 0.0f);
            var magenta = IDOMSolidColorBrush.createSolidCmyk(factory, 0.0f, 1.0f, 0.0f, 0.0f);
            var yellow = IDOMSolidColorBrush.createSolidCmyk(factory, 0.0f, 0.0f, 1.0f, 0.0f);
            var black = IDOMSolidColorBrush.createSolidCmyk(factory, 0.0f, 0.0f, 0.0f, 1.0f);

            // Create some spot color brushes
            var spotColorCyan = makeSeparationColor(factory, "LABCyan", new double[] { 62.0, -44.0, -50.0 });
            var spotColorMagenta = makeSeparationColor(factory, "LABMagenta", new double[] { 52.0, 81.0, -7.0 });
            var spotColorYellow = makeSeparationColor(factory, "LABYellow", new double[] { 95.0, -6.0, 96.0 });
            var spotColorBlack = makeSeparationColor(factory, "LABBlack", new double[] { 12.0, 2.0, 0.0 });
            var labCyan = IDOMSolidColorBrush.create(factory, spotColorCyan);
            var labMagenta = IDOMSolidColorBrush.create(factory, spotColorMagenta);
            var labYellow = IDOMSolidColorBrush.create(factory, spotColorYellow);
            var labBlack = IDOMSolidColorBrush.create(factory, spotColorBlack);

            // Create an 'All' spot color space, a color and a brush to draw with
            var spotColorAll = makeSeparationColor(factory, "All", new double[] { 1.0, 1.0, 1.0, 1.0 });
            var all = IDOMSolidColorBrush.create(factory, spotColorAll);

            // Create a 'Rot' spot color
            var spotColorRot = makeSeparationColor(factory, "Rot", new double[] { 0.0, 1.0, 1.0, 0.0 });
            var rot = IDOMSolidColorBrush.create(factory, spotColorRot);

            // Create a 'Blau' spot color
            var spotColorBlau = makeSeparationColor(factory, "Blau", new double[] { 1.0, 1.0, 0.0, 0.0 });
            var blau = IDOMSolidColorBrush.create(factory, spotColorBlau);

            // Create a 'Grun' spot color
            var spotColorGrun = makeSeparationColor(factory, "Grun", new double[] { 1.0, 0.0, 1.0, 0.0 });
            var grun = IDOMSolidColorBrush.create(factory, spotColorGrun);

            // Draw rows of shapes
            FPoint masterBoxSize = new FPoint(80, 80);
            FPoint origin = new FPoint(140, 75);

            drawRow(fixedPage, factory, origin, masterBoxSize, cyan, magenta, yellow, black, ShapeType.Box);
            origin.setY(origin.getY() + 200);
            drawRow(fixedPage, factory, origin, masterBoxSize, labCyan, labMagenta, labYellow, labBlack, ShapeType.Ellipse);
            origin.setY(origin.getY() + 200);
            drawRow(fixedPage, factory, origin, masterBoxSize, rot, grun, blau, all, ShapeType.Hexagon);
            origin.setY(origin.getY() + 200);
            drawRow(fixedPage, factory, origin, masterBoxSize, cyan, magenta, yellow, black, ShapeType.Polygon, 8, 22.5);
            origin.setY(origin.getY() + 200);
            drawRow(fixedPage, factory, origin, masterBoxSize, cyan, magenta, yellow, black, ShapeType.Target);

            var pdf = IPDFOutput.create(mako);
            pdf.writeAssembly(assembly, "test.pdf");
        }
        catch (Exception e) {
            System.out.println("Exception: " + e.getMessage());
        }
    }

    static void drawRow(IDOMFixedPage page, IEDLClassFactory factory, FPoint origin, FPoint masterBoxSize,
                        IDOMBrush cyan, IDOMBrush magenta, IDOMBrush yellow, IDOMBrush black,
                        ShapeType type) {
        drawRow(page, factory, origin, masterBoxSize, cyan, magenta, yellow, black, type, 0, 0.0);
    }

    static void drawRow(IDOMFixedPage page, IEDLClassFactory factory, FPoint origin, FPoint masterBoxSize,
                        IDOMBrush cyan, IDOMBrush magenta, IDOMBrush yellow, IDOMBrush black,
                        ShapeType type, int sides, double angle) {

        IDOMBrush[] brushes = new IDOMBrush[]{cyan, magenta, yellow, black};
        FPoint boxSize = new FPoint(masterBoxSize.getX(), masterBoxSize.getY());
        FPoint start = new FPoint(origin.getX(), origin.getY());

        for (IDOMBrush brush : brushes) {
            FRect strokeBox = new FRect(start.getX(), start.getY(), boxSize.getX(), boxSize.getY());
            var strokeGeom = createGeometry(factory, strokeBox, type, sides, angle);
            page.appendChild(IDOMPathNode.createStroked(factory, strokeGeom, brush));

            FRect fillBox = new FRect(strokeBox.getX(), strokeBox.getY() + boxSize.getY() * 1.1, boxSize.getX(), boxSize.getY());
            var fillGeom = createGeometry(factory, fillBox, type, sides, angle);
            page.appendChild(IDOMPathNode.createFilled(factory, fillGeom, brush));

            start.setX(start.getX() + boxSize.getX() * 1.1);
            boxSize.setX(boxSize.getX() * 1.25);
        }
    }

    static IDOMPathGeometry createGeometry(IEDLClassFactory factory, FRect box, ShapeType type, int sides, double angle) {
        switch (type) {
            case Ellipse:
                return IDOMPathGeometry.createEllipse(factory, box);
            case Hexagon:
                return IDOMPathGeometry.createPolygon(factory, box, 6, angle);
            case Polygon:
                return IDOMPathGeometry.createPolygon(factory, box, (short)sides, angle);
            case Target:
                return createTarget(factory, box, 8, angle);
            default:
                return IDOMPathGeometry.create(factory, box);
        }
    }

    static IDOMPathGeometry createTarget(IEDLClassFactory factory, FRect box, int nLines, double rotation) {
        double angle = Math.toRadians(rotation);
        double incr = 2.0 * Math.PI / nLines;
        double radiusX = box.getDX() / 2.0;
        double radiusY = box.getDY() / 2.0;
        double centerX = box.getX() + radiusX;
        double centerY = box.getY() + radiusY;

        var builder = IDOMPathGeometryBuilder.create(factory);
        FPoint center = new FPoint(centerX, centerY);
        builder.moveTo(center);

        for (int i = 0; i < nLines; i++) {
            double endX = centerX + radiusX * Math.cos(angle);
            double endY = centerY + radiusY * Math.sin(angle);
            builder.lineTo(new FPoint(endX, endY));
            builder.moveTo(center);
            angle += incr;
        }

        builder.close();
        return builder.createGeometry(factory, IDOMPathGeometry.eFillRule.eFRNonZero);
    }

    static IDOMColor makeSeparationColor(IEDLClassFactory factory, String name, double[] representation) {
        var colorant = new IDOMColorSpaceDeviceN.CColorantInfo(name, new CEDLVectDouble(representation));
        var colorants = new CEDLVectColorantInfo();
        colorants.append(colorant);

        var alternate = representation.length == 3 ? IDOMColorSpaceLAB.create(factory, 0.9642f, 1.0f, 0.8249f,
                0.0f, 0.0f, 0.0f, -128.0f, 127.0f, -128.0f, 127.0f) : IDOMColorSpaceDeviceCMYK.create(factory);

        var colorSpace = IDOMColorSpaceDeviceN.create(factory, colorants, alternate);
        return IDOMColor.create(factory, colorSpace, 1.0, 1.0);
    }
}

File: Python Projects/CreateShapes/spot_color_shapes.py
# -----------------------------------------------------------------------
# <copyright file="spot_color_shapes.py" company="Hybrid Software Helix Ltd">
#  Copyright (C) 2025 Hybrid Software Helix Ltd. All rights reserved.
# </copyright>
# <summary>
#  This example is provided on an "as is" basis and without warranty of any kind.
#  Hybrid Software Helix Ltd. does not warrant or make any representations
#  regarding the use or results of use of this example.
# </summary>
# -----------------------------------------------------------------------

import math
from jawsmakoIF_python import *


class ShapeType:
    Box = 0
    Ellipse = 1
    Hexagon = 2
    Polygon = 3
    Target = 4


def main():
    try:
        # Instantiate Mako
        jaws_mako = IJawsMako.create("")
        IJawsMako.enableAllFeatures(jaws_mako)
        factory = jaws_mako.getFactory()

        # Create the document and page structure
        assembly = IDocumentAssembly.create(jaws_mako)
        document = IDocument.create(jaws_mako)
        assembly.appendDocument(document)
        page = IPage.create(jaws_mako)
        document.appendPage(page)
        fixed_page = IDOMFixedPage.create(factory)
        page.setContent(fixed_page)

        # Create some process color brushes
        cyan = IDOMSolidColorBrush.createSolidCmyk(factory,1.0, 0.0, 0.0, 0.0)
        magenta = IDOMSolidColorBrush.createSolidCmyk(factory, 0.0, 1.0, 0.0, 0.0)
        yellow = IDOMSolidColorBrush.createSolidCmyk(factory, 0.0, 0.0, 1.0, 0.0)
        black = IDOMSolidColorBrush.createSolidCmyk(factory, 0.0, 0.0, 0.0, 1.0)

        # Create some spot color brushes
        spotColorCyan = make_separation_color(factory, "LABCyan", [ 62.0, -44.0, -50.0] )
        spotColorMagenta = make_separation_color(factory, "LABMagenta", [ 52.0, 81.0, -7.0 ])
        spotColorYellow = make_separation_color(factory, "LABYellow", [ 95.0, -6.0, 96.0 ])
        spotColorBlack = make_separation_color(factory, "LABBlack", [ 12.0, 2.0, 0.0 ])
        labCyan = IDOMSolidColorBrush.create(factory, spotColorCyan)
        labMagenta = IDOMSolidColorBrush.create(factory, spotColorMagenta)
        labYellow = IDOMSolidColorBrush.create(factory, spotColorYellow)
        labBlack = IDOMSolidColorBrush.create(factory, spotColorBlack)

         # Create an 'All' spot color space, a color and a brush to draw with
        spotColorAll = make_separation_color(factory, "All", [ 1.0, 1.0, 1.0, 1.0 ])
        all = IDOMSolidColorBrush.create(factory, spotColorAll)

         # Create a 'Rot' spot color
        spotColorRot = make_separation_color(factory, "Rot", [ 0.0, 1.0, 1.0, 0.0 ])
        rot = IDOMSolidColorBrush.create(factory, spotColorRot)

         # Create a 'Blau' spot color
        spotColorBlau = make_separation_color(factory, "Blau", [ 1.0, 1.0, 0.0, 0.0 ])
        blau = IDOMSolidColorBrush.create(factory, spotColorBlau)

        # Create a 'Grun' spot color
        spotColorGrun = make_separation_color(factory, "Grun", [ 1.0, 0.0, 1.0, 0.0 ])
        grun = IDOMSolidColorBrush.create(factory, spotColorGrun)

        # Draw rows of shapes
        master_box_size = FPoint(80, 80)
        origin = FPoint(140, 75)

        draw_row(fixed_page, factory, origin, master_box_size, cyan, magenta, yellow, black, ShapeType.Box)
        origin.y += 200
        draw_row(fixed_page, factory, origin, master_box_size, labCyan, labMagenta, labYellow, labBlack, ShapeType.Ellipse)
        origin.y += 200
        draw_row(fixed_page, factory, origin, master_box_size, rot, grun, blau, all, ShapeType.Hexagon)
        origin.y += 200
        draw_row(fixed_page, factory, origin, master_box_size, cyan, magenta, yellow, black,
                 ShapeType.Polygon, sides=8, angle=22.5)
        origin.y += 200
        draw_row(fixed_page, factory, origin, master_box_size, cyan, magenta, yellow, black, ShapeType.Target)

        # Write to PDF
        pdf = IPDFOutput.create(jaws_mako)
        pdf.writeAssembly(assembly, "test.pdf")

    except Exception as e:
        print("Exception:", str(e))


def draw_row(page, factory, origin, master_box_size, cyan, magenta, yellow, black, shape_type, sides=0, angle=0.0):
    brushes = [cyan, magenta, yellow, black]
    box_size = FPoint(master_box_size.x, master_box_size.y)
    start = FPoint(origin.x, origin.y)

    for brush in brushes:
        stroke_box = FRect(start.x, start.y, box_size.x, box_size.y)
        stroke_geom = create_geometry(factory, stroke_box, shape_type, sides, angle)
        page.appendChild(IDOMPathNode.createStroked(factory, stroke_geom, brush))

        fill_box = FRect(stroke_box.x, stroke_box.y + box_size.y * 1.1, box_size.x, box_size.y)
        fill_geom = create_geometry(factory, fill_box, shape_type, sides, angle)
        page.appendChild(IDOMPathNode.createFilled(factory, fill_geom, brush))

        start.x += box_size.x * 1.1
        box_size.x *= 1.25


def create_geometry(factory, box, shape_type, sides, angle):
    if shape_type == ShapeType.Ellipse:
        return IDOMPathGeometry.createEllipse(factory, box)
    elif shape_type == ShapeType.Hexagon:
        return IDOMPathGeometry.createPolygon(factory, box, 6, angle)
    elif shape_type == ShapeType.Polygon:
        return IDOMPathGeometry.createPolygon(factory, box, int(sides), angle)
    elif shape_type == ShapeType.Target:
        return create_target(factory, box, 8, angle)
    else:
        return IDOMPathGeometry.create(factory, box)


def create_target(factory, box, n_lines, rotation):
    angle = math.radians(rotation)
    incr = 2.0 * math.pi / n_lines
    radius_x = box.dX / 2.0
    radius_y = box.dY / 2.0
    center_x = box.x + radius_x
    center_y = box.y + radius_y

    builder = IDOMPathGeometryBuilder.create(factory)
    center = FPoint(center_x, center_y)
    builder.moveTo(center)

    for _ in range(n_lines):
        end_x = center_x + radius_x * math.cos(angle)
        end_y = center_y + radius_y * math.sin(angle)
        builder.lineTo(FPoint(end_x, end_y))
        builder.moveTo(center)
        angle += incr

    builder.close()
    return builder.createGeometry(factory, IDOMPathGeometry.eFRNonZero)


def make_separation_color(factory, name, representation):
    colorant = IDOMColorSpaceDeviceN.CColorantInfo(name, CEDLVectDouble(representation))
    colorants = CEDLVectColorantInfo()
    colorants.append(colorant)

    alternate = IDOMColorSpaceLAB.create(factory,
                                         0.9642, 1.0, 0.8249,
                                         0.0, 0.0, 0.0,
                                         -128.0, 127.0, -128.0, 127.0) if len(representation) == 3 else IDOMColorSpaceDeviceCMYK.create(factory)
    color_space = IDOMColorSpaceDeviceN.create(factory, colorants, alternate)
    return IDOMColor.create(factory, color_space, 1.0, 1.0)


if __name__ == "__main__":
    main()
